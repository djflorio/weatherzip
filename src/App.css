.App__fetching {
  padding: 30px 0;
  text-align: center;
}

.App__fetching-text {
  margin: 0;
  color: #fff;
  font-weight: 800;
}