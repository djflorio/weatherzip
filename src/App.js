// Third Party
import React, { Component } from 'react';
import axios from 'axios';
import uniqid from 'uniqid';

// Assets
import './assets/Animations.css';
import './App.css';

// Components
import Header from './components/header/Header';
import Search from './components/search/Search';
import AlertList from './components/alerts/Alerts';
import { HourGlass } from './components/spinners/Spinners';


class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      fetching: false,
      zip: "",
      country: "US",
      alerts: []
    };

    this.getWeather = this.getWeather.bind(this);
    this.setValue = this.setValue.bind(this);
    this.addAlert = this.addAlert.bind(this);
    this.removeAlert = this.removeAlert.bind(this);
  }

  getWeather(e) {
    e.preventDefault();
    const zip = this.state.zip;

    if (zip === "") {
      this.addAlert("Zip code required");
      return;
    }

    const country = this.state.country;
    const key = "396295ac4fa8d1ebfbd1229f7809fe56";
    const url = `https://api.openweathermap.org/data/2.5/forecast?zip=${zip},${country}&units=imperial&appid=${key}`;

    this.setState({ fetching: true });
    axios.get(url)
    .then(res => {
      this.setState({ fetching: false });
      console.log(res);
    })
    .catch(err => {
      this.setState({ fetching: false });
      if (err.response.status === 404) {
        this.addAlert("No weather data found for that location");
      }
    });
  }

  setValue(e) {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  addAlert(message) {
    const newAlerts = this.state.alerts.slice();
    const alert = {
      id: uniqid(),
      message: message
    };
    newAlerts.push(alert);
    this.setState({
      alerts: newAlerts
    });
  }

  removeAlert(id) {
    const newAlerts = this.state.alerts.slice();
    for (let i = 0; i < newAlerts.length; i++) {
      if (newAlerts[i].id === id) {
        newAlerts.splice(i, 1);
        break;
      }
    }
    this.setState({
      alerts: newAlerts
    });
  }

  render() {
    return (
      <div className="App">
        <AlertList
          alerts={this.state.alerts}
          onRemove={this.removeAlert}
        />
        <Header />
        {
          !this.state.fetching &&
          <Search
            country={this.state.country}
            zip={this.state.zip}
            onChange={this.setValue}
            onSubmit={this.getWeather}
          />
        }
        {
          this.state.fetching &&
          <div className="App__fetching">
            <HourGlass />
            <p className="App__fetching-text">Getting your forecast</p>
          </div>
        }
      </div>
    );
  }
}

export default App;
