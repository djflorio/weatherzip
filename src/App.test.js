import App from './App';

function setupShallow() {
  return shallow(<App />);
}

it('renders correctly', () => {
  const wrapper = setupShallow();
  expect(wrapper).toMatchSnapshot();
});

describe('setUnits', () => {
  it('sets units in App state', () => {
    const wrapper = setupShallow();
    wrapper.instance().setUnits("metric");
    expect(wrapper.state().units).toEqual("metric");
  });
});

describe('resetSearch', () => {
  it('resets the search in App state', () => {
    const wrapper = setupShallow();
    wrapper.instance().resetSearch();
    expect(wrapper.state()).toEqual(
      expect.objectContaining({
        fetching: false,
        data: []
      })
    );
  });
});

describe('removeAlert', () => {
  it('removes an alert from the App state', () => {
    const wrapper = setupShallow();
    wrapper.state().alerts = [
      { id: "1", message: "Test" },
      { id: "2", message: "Test" }
    ];
    wrapper.instance().removeAlert("2");
    expect(wrapper.state().alerts.length).toEqual(1);
  });
});

describe('addAlert', () => {
  it('adds an alert to the App state', () => {
    const wrapper = setupShallow();
    wrapper.instance().addAlert("Hello");
    const alerts = wrapper.state().alerts;
    expect(alerts.length).toEqual(1);
    expect(alerts[0]).toEqual({
      id: expect.anything(),
      message: "Hello"
    });
  });
});

/*describe('getWeather', () => {
  it('should add alert to state if no zip in state', () => {
    const wrapper = setupShallow();
    wrapper.state().zip = "";
    wrapper.instance().getWeather(wrapper.simulate("change"));
    expect(wrapper.state().alerts.length).toEqual(1);
  });
});*/