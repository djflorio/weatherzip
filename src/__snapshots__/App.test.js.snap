// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <AlertList
    alerts={Array []}
    onRemove={[Function]}
  />
  <Header />
  <Search
    country="US"
    onChange={[Function]}
    onSubmit={[Function]}
    onUnitChange={[Function]}
    units="imperial"
    zip=""
  />
  <Footer />
</div>
`;
